drivers/counter/counter_smartbond_timer.c

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright (c) 2022 Renesas Electronics Corporation and/or its affiliates
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_smartbond_timer
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/irq.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+#include <DA1469xAB.h>
+
+LOG_MODULE_REGISTER(counter_timer, CONFIG_COUNTER_LOG_LEVEL);
+
+#define LP_CLK_OSC_RC32K   0
+#define LP_CLK_OSC_RCX	   1
+#define LP_CLK_OSC_XTAL32K 2
+
+#define TIMER_TOP_VALUE 0xFFFFFF
+
+struct counter_smartbond_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+	uint32_t guard_period;
+	uint32_t freq;
+};
+
+struct counter_smartbond_ch_data {
+	counter_alarm_callback_t callback;
+	void *user_data;
+};
+
+struct counter_smartbond_config {
+	struct counter_config_info info;
+	/* Register set for timer */
+	TIMER2_Type *timer;
+	uint8_t prescaler;
+	/* Timer driven by DIVn if 1 or lp_clk if 0 */
+	uint8_t clock_src_divn;
+	uint8_t irqn;
+	void (*irq_config_func)(const struct device *dev);
+
+	LOG_INSTANCE_PTR_DECLARE(log);
+};
+
+static int counter_smartbond_start(const struct device *dev)
+{
+	const struct counter_smartbond_config *config = dev->config;
+	TIMER2_Type *timer = config->timer;
+
+	/* enable counter in free running mode */
+	timer->TIMER2_CTRL_REG |= TIMER_TIMER_CTRL_REG_TIM_CLK_EN_Msk |
+				  TIMER_TIMER_CTRL_REG_TIM_EN_Msk |
+				  TIMER_TIMER_CTRL_REG_TIM_FREE_RUN_MODE_EN_Msk;
+
+	return 0;
+}
+
+static int counter_smartbond_stop(const struct device *dev)
+{
+	const struct counter_smartbond_config *config = dev->config;
+	struct counter_smartbond_data *data = dev->data;
+	TIMER2_Type *timer = config->timer;
+
+	/* disable counter */
+	timer->TIMER2_CTRL_REG &=
+		~(TIMER2_TIMER2_CTRL_REG_TIM_EN_Msk | TIMER2_TIMER2_CTRL_REG_TIM_IRQ_EN_Msk);
+	data->callback = NULL;
+
+	return 0;
+}
+
+static uint32_t counter_smartbond_get_top_value(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return TIMER_TOP_VALUE;
+}
+
+static uint32_t counter_smartbond_read(const struct device *dev)
+{
+	const struct counter_smartbond_config *config = dev->config;
+	TIMER2_Type *timer = config->timer;
+
+	return timer->TIMER2_TIMER_VAL_REG;
+}
+
+static int counter_smartbond_get_value(const struct device *dev, uint32_t *ticks)
+{
+	*ticks = counter_smartbond_read(dev);
+
+	return 0;
+}
+
+static int counter_smartbond_set_alarm(const struct device *dev, uint8_t chan,
+				       const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_smartbond_config *config = dev->config;
+	struct counter_smartbond_data *data = dev->data;
+	TIMER2_Type *timer = config->timer;
+
+	__IOM uint32_t *timer_clear_irq_reg = ((TIMER_Type *)timer) == TIMER
+						      ? &((TIMER_Type *)timer)->TIMER_CLEAR_IRQ_REG
+						      : &timer->TIMER2_CLEAR_IRQ_REG;
```

Comment:
this is "calculated" more than once in driver (`counter_smartbond_irq_handler`) so perhaps make it a helper func

Reply:
Refactored in one function